ksql-parser/src/main/java/io/confluent/ksql/parser/tree/PrimitiveType.java

```diff
@@ -15,51 +15,142 @@
 
 package io.confluent.ksql.parser.tree;
 
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.google.errorprone.annotations.Immutable;
 import io.confluent.ksql.util.KsqlException;
+
+import java.util.List;
 import java.util.Objects;
 import java.util.Optional;
+import java.util.function.Function;
+import javax.annotation.Nullable;
 
 @Immutable
 public final class PrimitiveType extends Type {
 
-  private static final ImmutableMap<SqlType, PrimitiveType> TYPES =
-      ImmutableMap.<SqlType, PrimitiveType>builder()
-      .put(SqlType.BOOLEAN, new PrimitiveType(SqlType.BOOLEAN))
-      .put(SqlType.INTEGER, new PrimitiveType(SqlType.INTEGER))
-      .put(SqlType.BIGINT,  new PrimitiveType(SqlType.BIGINT))
-      .put(SqlType.DOUBLE,  new PrimitiveType(SqlType.DOUBLE))
-      .put(SqlType.STRING,  new PrimitiveType(SqlType.STRING))
-      .build();
+  /* List of parameters for the type (i.e. DECIMAL(6,2)). */
+  final ImmutableList<Integer> typeParameters;
 
+  private static final ImmutableMap<SqlType, Function<ImmutableList<Integer>, PrimitiveType>>
+      TYPES = ImmutableMap.<SqlType, Function<ImmutableList<Integer>, PrimitiveType>>builder()
+      .put(SqlType.BOOLEAN, p -> new PrimitiveType(SqlType.BOOLEAN))
+      .put(SqlType.INTEGER, p -> new PrimitiveType(SqlType.INTEGER))
+      .put(SqlType.BIGINT,  p -> new PrimitiveType(SqlType.BIGINT))
+      .put(SqlType.DOUBLE,  p -> new PrimitiveType(SqlType.DOUBLE))
+      .put(SqlType.STRING,  p -> new PrimitiveType(SqlType.STRING))
+      .put(SqlType.DECIMAL, p -> new PrimitiveType(SqlType.DECIMAL, p))
+      .build();
 
   public static PrimitiveType of(final String typeName) {
+    return of(typeName, ImmutableList.of());
+  }
+
+  public static PrimitiveType of(
+      final String typeName,
+      @Nullable final List<Integer> typeParameters
+  ) {
     switch (typeName.toUpperCase()) {
       case "INT":
         return PrimitiveType.of(SqlType.INTEGER);
       case "VARCHAR":
         return PrimitiveType.of(SqlType.STRING);
+      case "DEC": // SQL standard states that DEC is similar to DECIMAL
+        return PrimitiveType.of(SqlType.DECIMAL, typeParameters);
       default:
         try {
           final SqlType sqlType = SqlType.valueOf(typeName.toUpperCase());
-          return PrimitiveType.of(sqlType);
+          return PrimitiveType.of(sqlType, typeParameters);
         } catch (final IllegalArgumentException e) {
           throw new KsqlException("Unknown primitive type: " + typeName, e);
         }
     }
   }
 
   public static PrimitiveType of(final SqlType sqlType) {
-    final PrimitiveType primitive = TYPES.get(Objects.requireNonNull(sqlType, "sqlType"));
-    if (primitive == null) {
+    return function(sqlType).apply(ImmutableList.of());
+  }
+
+  public static PrimitiveType of(
+      final SqlType sqlType,
+      @Nullable final List<Integer> sqlTypeParameters
```

Comment:
Done. Btw, I still need to use the @Nullable annotation on one of the of() methods because the `ParserModelTest` was failing if null was passed as a 2nd parameter just because I was not throwing a NPE.

Reply:
Yes, but as I said - avoid the use of nulls - much better to use `Optional`. `@Nullbable` is not tested by all compilers / IDEs and isn't as clear as `Optional`.

Reply:
Also, change your code so that is _does_ throw a NPE if null is passed.